feat: Focus order for Left sidebar, context menu, conversation list and footer menu(ACC-267)

arjita-mitra opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/ACC-267" title="ACC-267" target="_blank"><img alt="Subtask" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10816?size=medium" />ACC-267</a>  9.2.4.3 Focus order for Left sidebar - Preferences, conversation list and menu
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - Focus order for Left sidebar, context menu, conversation list and footer menu
  - 
- The **PR Description**
  - The navigation sequence sketches can be found on here [Accessibility Quick-Wins (Web/Desktop)](https://www.figma.com/file/OrP9jAJdfttfvsm3nxnihP/Accessibility-Quick-Wins-(Web%2FDesktop)?node-id=0%3A1)

**Important changes:**

The navigation for left side bar via tab will be in the order
Preferences icon>Status change>Conversation list(first conversation)>Bottom Menu (Contacts>Conversations>Folders)

Within the conversation list, user will navigate via arrow keys between different conversations. 

To open the context menu user needs to press right arrow key and enter. Once the menu is open the first menu item gets the focus. Use arrow keys(up/down) to focus on prev/next menu item. selection of menu item(hit enter) will perform the action, close the menu and focus back to the event generator i.e context menu button.

Shift+tab brings the focus back in the order.

# What's new in this PR?

### Issues

_Left side bar is not keyboard accessible.

### Solutions

- implement the focus order as mentioned in [ACC-267](https://wearezeta.atlassian.net/browse/ACC-267)

### Testing

#### Test Coverage (Optional)

- use screen reader and keyboard for testing.

#### How to Test

- use screen reader and keyboard for testing.

### Attachments (Optional)
![Uploading Screenshot 2022-09-29 at 11.10.43.png…]()

----
----
##### References
1. https://wearezeta.atlassian.net/browse/ACC-267
2. https://www.figma.com/file/OrP9jAJdfttfvsm3nxnihP/Accessibility-Quick-Wins-(Web%2FDesktop)?node-id=0%3A1
3. https://www.freecodecamp.org/news/html-roving-tabindex-attribute-explained-with-examples/

arjita-mitra committed:
feat: Focus order for Left sidebar, Preferences, conversation list and footer menu(ACC-267)

arjita-mitra committed:
feat: sidebar folder menu focus order
shift+tab focus order(ACC-267)

arjita-mitra committed:
Merge branch 'dev' of https://github.com/wireapp/wire-webapp into feature/ACC-267

atomrc commented on src/script/hooks/useRoveFocus.ts:
```diff
@@ -0,0 +1,44 @@
+import {isKey, isTabKey, KEY} from 'Util/KeyboardUtil';
+import type {KeyboardEvent as ReactKeyboardEvent} from 'react';
+
+function useRoveFocus(size: number) {
```
I'm not sure I understand what is `roveFocus` can you provide a JS doc explanation of what this does, that would be nice :)

V-Gira replied:
roving tabindex is when you have custom keyboard navigation afaik, in that case, navigating the sidebar with arrowkeys.

arjita-mitra replied:
@atomrc this one explained well https://www.freecodecamp.org/news/html-roving-tabindex-attribute-explained-with-examples/

atomrc replied:
Please link this page in the code so that future dev will understand what is it about ;)

arjita-mitra committed:
feat: make archieve list accessible
user name with/without clickable should be keyboard accessible
resolve typescript error
tab key press while menu is open should close the menu(ACC-267)

arjita-mitra committed:
fix: typescript error(ACC-267)

arjita-mitra committed:
fix: lint errors(ACC-267)

atomrc commented on src/script/components/list/ConversationListCell.tsx:
```diff
@@ -35,17 +35,22 @@ import Avatar from 'Components/Avatar';
   showJoinButton: boolean;
+  index: number;
+  focus: boolean;
+  showFocus: boolean;
```
Can you document, in the code, what is the difference between `focus` and `showFocus` it's not quite clear. 

you can use `jsdoc` style comment so that we get the info on our IDE directly :) 

```suggestion
  /** explain here what the `focus` property does */
  focus: boolean;
  
  /** explain here what the `showFocus` property does */
  showFocus: boolean;
```

arjita-mitra committed:
fix: rename focus and showfocus property
add roving tabindex documentation link(ACC-267)
